### PR TITLE
Add moe loss normalization for RL SFT

### PR DESCRIPTION
**Pull request by pthombre**
# What does this PR do ?
This change adds a separate loss scaling function specifically for MoE (Mixture of Experts) auxiliary losses during RL SFT training.                                                                                                                        
                                                                                                                                                                                                                                                              
  Problem: Previously, MoE auxiliary losses used the same grad_scale_func as the main model loss. In RL SFT scenarios, you may need to scale the MoE auxiliary loss differently from the primary loss (e.g., the main loss uses a dynamic scaler tied to RL   
  rewards, but MoE load-balancing losses need their own scaling).
                                                                                                                                                                                                                                                              
  What changed:   
  - A new optional config field moe_grad_scale_func is added to ModelParallelConfig
  - In the forward step, MoE loss scaling now checks for moe_grad_scale_func first, falls back to grad_scale_func if not set, and defaults to 1 otherwise                                                                                                     
  - Note that moe_grad_scale_func takes no arguments (returns the scale directly), unlike grad_scale_func which takes a tensor input — this allows a simpler interface where the caller pre-computes the scale
                                                                                                                                                                                                                                                              
  Impact: Existing code is unaffected since moe_grad_scale_func defaults to None, preserving the current fallback behavior. 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Contribution process

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

File: megatron/core/pipeline_parallel/schedules.py [hunk 1]
```diff
@@ -289,13 +289,16 @@ def forward_step_calc_loss(
...
+        # grad_scale_func (fallback), or default to 1.
         device = get_tensor_device(output_tensor)
-        loss_scale = (
-            config.grad_scale_func(torch.ones(1, device=device))
-            if config.grad_scale_func is not None
-            else torch.ones(1, device=device)
-        )
+        moe_grad_scale_func = getattr(config, 'moe_grad_scale_func', None)
+        if moe_grad_scale_func is not None:
+            loss_scale = moe_grad_scale_func()
```
Comment (yashaswikarnati) [inline megatron/core/pipeline_parallel/schedules.py:323]: Would this return a constant value or is this dependent on the microbatch ? I wonder if this can cause issues with scale thats globally stored in  MoEAuxLossAutoScaler.

Comment (yashaswikarnati) [reply megatron/core/pipeline_parallel/schedules.py:323]: Like in interleaved 1F1B, the scale computed in one forward pass for a microbatch could be accidentally used for next immediate backward, which might be for a different backward

**Review by pthombre — COMMENTED** [review]
This is a valid callout. In this particular case moe_grad_scale_func returns a constant per global batch, not a per-microbatch value, so we should be okay even in the interleaved 1F1B case

**Review by yashaswikarnati — APPROVED** [review]

**Comment by yashaswikarnati** [comment]
> This is a valid callout. In this particular case moe_grad_scale_func returns a constant per global batch, not a per-microbatch value, so we should be okay even in the interleaved 1F1B case

thanks! Just a nit; would be still great to add a gaurd this incase someone passes a function that returns a fixed scalar and not silently cause wrong behaviour.

**Review by jaredcasper — APPROVED** [review]
